deps/chakrashim/src/jsrtcontextshim.cc

```diff
@@ -368,11 +368,34 @@ bool ContextShim::InitializeProxyOfGlobal() {
   });
 }
 
-bool ContextShim::EnsureInitialized() {
+
+bool ContextShim::InitializeCurrentContextShim() {
   if (initialized) {
     return true;
   }
 
+  JsContextRef oldContext;
+  if (JsGetCurrentContext(&oldContext) != JsNoError) {
+    return false;
+  }
+  JsContextRef thisContext = GetContextRef();
+  if (JsSetCurrentContext(thisContext) != JsNoError) {
+    return false;
+  }
+  if (!EnsureInitialized()) {
+    Fatal("Failed to initialize context");
+  }
+  if (JsSetCurrentContext(oldContext) != JsNoError) {
+    return false;
+  }
+  return true;
+}
+
+bool ContextShim::EnsureInitialized() {
+  if (initialized) {
+    return true;
+  }
+  initializing = true;
```

Comment:
I initially implemented mostly the way you suggested. However I was thinking of the scenario **in future** where we are in middle of `EnsureInitialized` and it calls getter on an object that has not yet initialized. Since `initialized = true` is at the beginning, the getter would think that everything is initialized and would return an uninitialized object. Currently it never happen because we have made sure that we initialize objects before they are used in initializing other objects.
To make code cleaner and ensure that we never use objects in initialization code without initializing, I have added an assert in getters to ensure that objects returned are never `JS_INVALID_REFERENCE`. With that we don't need `initializing` variable anymore.